DESIGN.md

```diff
@@ -0,0 +1,98 @@
+SSH is a multiplexed, packet-oriented protocol.
+
+This is different from SSL which is stream-oriented.
+And HTTP which is also stream-oriented.
+HTTP/2 is a stream-oriented, multiplexed protocol: it multiplexes/demultiplexes a number of streams into one.
+
+Because SSH is a packet-oriented, we can't use the [Pipe](https://docs.microsoft.com/en-us/dotnet/api/system.io.pipelines.pipe?view=dotnet-plat-ext-2.1) class after we've decrypted the packets.
+The `Pipe` class represents a stream buffer, and storing the packets into it would make us lose the packet boundaries.
+
+for buffering packets, we can make use of the [Channel](https://docs.microsoft.com/en-us/dotnet/api/system.threading.channels.channel?view=dotnet-plat-ext-2.1) class.
+
+As a type to represent packets, we want something that works wel together with the `Pipe` related types, like [Memory](https://docs.microsoft.com/en-us/dotnet/api/system.memory-1?view=netstandard-2.1), [ReadOnlySequence<T>](https://docs.microsoft.com/en-us/dotnet/api/system.buffers.readonlysequence-1?view=netstandard-2.1), and [SequenceReader<T>](https://docs.microsoft.com/en-us/dotnet/api/system.buffers.sequencereader-1?view=netcore-3.0).
+
+We want to pool the packet type so it can be re-used by the `SshConnection`.
+Also, if an instance was parsed, we may want to trim it and allow passing it to the user.
+
+So our packet should be some type of _mutable_ counterpart of `ReadOnlySequence`.
+```cs
+class Sequence
```

Comment:
Should this implement `IBufferWriter<byte>`?

Reply:
Also see https://github.com/AArnott/Nerdbank.Streams/blob/master/src/Nerdbank.Streams/Sequence%601.cs

Reply:
> Should this implement IBufferWriter<byte>?

It meets the contract.

> Also see https://github.com/AArnott/Nerdbank.Streams/blob/master/src/Nerdbank.Streams/Sequence%601.cs

This is similar to that type, but tied to `SequencePool` to provide pooling of `Memory`, `Segments`, and the `Sequences` themselves.